server/src/coverage/KcovCoverageTool.cpp

```diff
@@ -0,0 +1,131 @@
+#include "utils/JsonUtils.h"
+#include "utils/MakefileUtils.h"
+#include "utils/StringUtils.h"
+#include "utils/path/FileSystemPath.h"
+#include "pugixml.hpp"
+#include <regex>
+
+#include "loguru.h"
+
+using Coverage::CoverageMap;
+using Coverage::FileCoverage;
+
+
+KcovCoverageTool::KcovCoverageTool(utbot::ProjectContext projectContext,
+                                   ProgressWriter const *progressWriter)
+        : CoverageTool(progressWriter), projectContext(projectContext) {
+}
+
+std::vector<BuildRunCommand>
+KcovCoverageTool::getBuildRunCommands(const std::vector<UnitTest> &testsToLaunch, bool withCoverage) {
+    ExecUtils::throwIfCancelled();
+
+    std::vector<BuildRunCommand> result;
+    ExecUtils::doWorkWithProgress(
+            testsToLaunch, progressWriter, "Collecting build and run commands",
+            [&](UnitTest const &testToLaunch) {
+                auto makefile = Paths::getMakefilePathFromSourceFilePath(
+                        projectContext,
+                        Paths::testPathToSourcePath(projectContext, testToLaunch.testFilePath));
+                auto gtestFlags = getTestFilter(testToLaunch);
+                auto buildCommand =
+                        MakefileUtils::makefileCommand(projectContext, makefile, "build", gtestFlags);
+                auto runCommand =
+                        MakefileUtils::makefileCommand(projectContext, makefile, "run", gtestFlags);
+                result.push_back({testToLaunch, buildCommand, runCommand});
+            });
+    return result;
+}
+
+std::vector<ShellExecTask>
+KcovCoverageTool::getCoverageCommands(const std::vector<UnitTest> &testsToLaunch) {
+    return {};
+}
+
+
+static void addLine(uint32_t lineNumber, bool covered, FileCoverage &fileCoverage) {
+    assert(lineNumber > 0);
+    if (covered) {
+        fileCoverage.fullCoverageLines.insert({lineNumber});
+    } else {
+        fileCoverage.noCoverageLines.insert({lineNumber});
+    }
+}
+
+static void setLineNumbers(pugi::xml_node lines, FileCoverage &fileCoverage) {
+    for (pugi::xml_node line: lines.children("line")) {
+        uint32_t lineNumber = line.attribute("number").as_int();
+        bool covered = line.attribute("hits").as_int() > 0;
+        addLine(lineNumber, covered, fileCoverage);
+    }
+}
+
+Coverage::CoverageMap KcovCoverageTool::getCoverageInfo() const {
+    ExecUtils::throwIfCancelled();
+
+    CoverageMap coverageMap;
+
+    auto coverageReportDirPath = Paths::getKcovReportDir(projectContext);
+    if (!fs::exists(coverageReportDirPath)) {
+        std::string message = "Couldn't find coverage directory at " + coverageReportDirPath.string();
+        LOG_S(ERROR) << message;
+        throw CoverageGenerationException(message);
+    }
+    LOG_S(INFO) << "Reading coverage file";
+
+
+    std::string pathToReport = "";
+    for (const auto &entry: fs::directory_iterator(coverageReportDirPath.string())) {
+        std::regex reg("(calc.).*$");
```

Comment:
You can avoid building `std::regex` object every single time and construct the corresponding object just once per program, say, declaring it as a global const static variable